backend/module/chaos/serverexperimentation/rtds/envoy/envoy.go

```diff
@@ -0,0 +1,237 @@
+// final configuration.
+type EnvoyConfig struct {
+	configModifiers []func(*bootstrap.Bootstrap) *bootstrap.Bootstrap
+
+	finalConfig *bootstrap.Bootstrap
+}
+
+// Generate generates the final configuration by applying all the config modifiers.
+func (e *EnvoyConfig) Generate() (string, error) {
+	if e.finalConfig == nil {
+		b := &bootstrap.Bootstrap{}
+
+		err := unmarshalYaml(b, baseConfig)
+		if err != nil {
+			return "", err
+		}
+
+		for _, m := range e.configModifiers {
+			b = m(b)
+		}
+
+		e.finalConfig = b
+	}
+
+	m := jsonpb.Marshaler{}
+	out := bytes.NewBuffer([]byte{})
+	err := m.Marshal(out, e.finalConfig)
+
+	return out.String(), err
+}
+
+// NewEnvoyConfig creates a new Envoy config builder, using a sensible default configuration that allows
+// the provided EnvoyHandle to interact with the underlying Envoy instance.
+func NewEnvoyConfig() *EnvoyConfig {
+	return &EnvoyConfig{}
+}
+
+// AddRuntimeLayer adds a single runtime layer to the bootstrap.
+func (e *EnvoyConfig) AddRuntimeLayer(input string) error {
+	runtimeLayer := &bootstrap.RuntimeLayer{}
+
+	err := unmarshalYaml(runtimeLayer, input)
+	if err != nil {
+		return err
+	}
+
+	e.configModifiers = append(e.configModifiers, func(b *bootstrap.Bootstrap) *bootstrap.Bootstrap {
+		if b.LayeredRuntime == nil {
+			b.LayeredRuntime = &bootstrap.LayeredRuntime{}
+		}
+		b.LayeredRuntime.Layers = append(b.LayeredRuntime.Layers, runtimeLayer)
+
+		return b
+	})
+
+	return nil
+}
+
+// AddCluster adds a cluster to the list of static clusters.
+func (e *EnvoyConfig) AddCluster(input string) error {
+	cluster := &cluster.Cluster{}
+
+	err := unmarshalYaml(cluster, input)
+	if err != nil {
+		return err
+	}
+
+	e.configModifiers = append(e.configModifiers, func(b *bootstrap.Bootstrap) *bootstrap.Bootstrap {
+		b.StaticResources.Clusters = append(b.StaticResources.Clusters, cluster)
+
+		return b
+	})
+
+	return nil
+}
+
+// AddHTTPFilter adds a HTTP filter in front of the list of HTTP filters for the default listener.
+func (e *EnvoyConfig) AddHTTPFilter(input string) error {
+	filter := &hcm.HttpFilter{}
+	err := unmarshalYaml(filter, input)
+	if err != nil {
+		return err
+	}
+
+	e.configModifiers = append(e.configModifiers, func(b *bootstrap.Bootstrap) *bootstrap.Bootstrap {
+		h := &hcm.HttpConnectionManager{}
+		// TODO(snowp): Have config modifiers return (Bootstrap, error) so we can propagate errors.
+		_ = b.StaticResources.Listeners[0].FilterChains[0].Filters[0].GetTypedConfig().UnmarshalTo(h)
+
+		h.HttpFilters = append([]*hcm.HttpFilter{filter}, h.HttpFilters...)
+
+		a, _ := ptypes.MarshalAny(h)
+
+		b.StaticResources.Listeners[0].FilterChains[0].Filters[0].ConfigType = &listener.Filter_TypedConfig{
+			TypedConfig: a,
+		}
+
+		return b
+	})
+
+	return nil
+}
+
+// Helper to convert an input yaml into a typed Protobuf message.
+func unmarshalYaml(m proto.Message, input string) error {
```

Comment:
helpers exist for this already in [backend/mock/api/api.go](https://github.com/lyft/clutch/blob/4a8a318e2916a01dd1d6c6b5ff5ab0285a4b324e/backend/mock/api/api.go)